### DISCUSSION TITLE
SyntaxError: 'persist' is not a valid YOLO argument.

### DISCUSSION THREAD

**Discussion by Wouter-Bekker-AI** [ultralytics/ultralytics#3222] [Q&A] [answered]
When using tracking as shown in the tracking readme: 
If you want to use the tracker with a folder of images or when you loop on the video frames, you should use the persist parameter to tell the model that these frames are related to each other so the IDs will be fixed for the same objects. Otherwise, the IDs will be different in each frame because in each loop, the model creates a new object for tracking, but the persist parameter makes it use the same object for tracking.

import cv2
from ultralytics import YOLO

cap = cv2.VideoCapture("video.mp4")
model = YOLO("yolov8n.pt")
while True:
    ret, frame = cap.read()
    if not ret:
        break
    results = model.track(frame, persist=True)
    boxes = results[0].boxes.xyxy.cpu().numpy().astype(int)
    ids = results[0].boxes.id.cpu().numpy().astype(int)
    for box, id in zip(boxes, ids):
        cv2.rectangle(frame, (box[0], box[1]), (box[2], box[3]), (0, 255, 0), 2)
        cv2.putText(
            frame,
            f"Id {id}",
            (box[0], box[1]),
            cv2.FONT_HERSHEY_SIMPLEX,
            1,
            (0, 0, 255),
            2,
        )
    cv2.imshow("frame", frame)
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break
I get this error: 
...
SyntaxError: 'persist' is not a valid YOLO argument. 

    Arguments received: ['yolo']. Note that Ultralytics 'yolo' commands use the following syntax:
...

I want to use the tracker to track every frame of a rtsp feed

**Comment by ritesh27gole** [comment]
I am getting the same issue. Was it solved for you @Wouter-Bekker-AI ?

**Reply by Wouter-Bekker-AI** [reply]
No not yet. Does anyone have an idea what might cause the problem?

**Reply by pderrenger** [reply]
@Wouter-Bekker-AI hello! It seems there might be a misunderstanding regarding the use of the `persist` argument in the tracking mode of YOLOv8. The `persist` argument is not a recognized parameter in the current YOLOv8 tracking API.

To perform tracking on a video or a stream of frames where you want to maintain consistent object IDs across frames, you should initialize the tracker once outside of your frame reading loop and then call the `track` method on each frame without the need for a `persist` argument. The tracker will automatically maintain object IDs across the frames.

Here's an updated version of your code snippet without the `persist` argument:

```python
import cv2
from ultralytics import YOLO

cap = cv2.VideoCapture("video.mp4")
model = YOLO("yolov8n.pt")  # Initialize the model with the tracking weights

while True:
    ret, frame = cap.read()
    if not ret:
        break
    results = model.track(frame)  # Call track on each frame
    boxes = results.boxes.xyxy.cpu().numpy().astype(int)
    ids = results.boxes.id.cpu().numpy().astype(int)
    for box, id in zip(boxes, ids):
        cv2.rectangle(frame, (box[0], box[1]), (box[2], box[3]), (0, 255, 0), 2)
        cv2.putText(
            frame,
            f"Id {id}",
            (box[0], box[1]),
            cv2.FONT_HERSHEY_SIMPLEX,
            1,
            (0, 0, 255),
            2,
        )
    cv2.imshow("frame", frame)
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

cap.release()
cv2.destroyAllWindows()
```

Make sure you are using the latest version of the Ultralytics YOLOv8 repository, and if you encounter any further issues, please refer to the [Track documentation](https://docs.ultralytics.com/modes/track) for more detailed guidance on how to use the tracking functionality. If the issue persists, feel free to open an issue on the repository with the details of the problem you're facing.

**Comment by djoshi1000** [comment, marked as answer]
Update your YOLOv8. It worked for me after updating YOLOv8 on my computer @Wouter-Bekker-AI @ritesh27gole

**Reply by ritesh27gole** [reply]
Thanks a lot. I wonder why it did not work when I updated it last time!

**Reply by Wouter-Bekker-AI** [reply]
It works for me as well now, thanks.

**Reply by Mieczmik** [reply]
how to update yolo?

**Reply by pderrenger** [reply]
@Mieczmik to update your YOLOv8 installation, you should follow these steps:

1. Open your command line interface (CLI).
2. Activate the Python environment where YOLOv8 is installed, if you're using a virtual environment.
3. Use the `pip` command to install the latest version of the `ultralytics` package, which includes YOLOv8. You can do this with the following command:

```bash
pip install --upgrade ultralytics
```

This command will download and install the latest version of the `ultralytics` package from the Python Package Index (PyPI), which includes the most recent updates and bug fixes for YOLOv8.

Remember to check the [Ultralytics Docs](https://docs.ultralytics.com) for any additional steps or changes that might be required after updating. If you encounter any issues or have further questions, feel free to reach out! 😊👍
